Cache scope (#89)
buildpacks/rfcs: text/0000-cache-types.md

- CNB Issue: (leave blank)
- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)

# Summary
[summary]: #summary

Allow for buildpacks to specify what cache layers can be shared across applications.

# Motivation
[motivation]: #motivation

In order to provide a better experience, platforms such as `pack` could provide a way to share cache across multiple applications. In particular, this is a frequently requested feature ([pack#340](https://github.com/buildpacks/pack/issues/340), [pack#652](https://github.com/buildpacks/pack/issues/652), [slack](https://buildpacks.slack.com/archives/C94UJCNV6/p1588718688412300), [slack](https://buildpacks.slack.com/archives/CD61YAG69/p1586370336098700)) for similar language family apps.

Use Case:
```
Given I build a Java app A
When I build a Java app B
Then I expect similar maven dependencies not to have to be downloaded again
```

In order to solve this use case in a safe manner, the platform needs to know what cache it can share with other applications versus what cache is specific to a single application. This information is primarily known by the buildpacks.

# What it is
[what-it-is]: #what-it-is

> Buildpacks will provide a mechanism to denote a cached layer as something that can be shared by associating specifing a scope of "shared".

Reviewer: ```suggestion
Buildpacks will provide a mechanism to denote a cached layer as something that can be shared by specifying a scope of "shared".
```

- Target persona: buildpack author

There will be two defined scopes: 

- `app` - This cached layer is specific to the application and SHOULD NOT be shared with any other applications.

> - `shared` - This cached layer is specific to the buildpack and MAY be shared across multiple applications.

Reviewer 2: If there are only two allowed scopes should we just have a `shared` bool?

# How it Works
[how-it-works]: #how-it-works

#### Platform Implementation

> Platforms can decide how to _share_ the "shared" cached layers based on their specific use cases.

Reviewer: Could a platform opt to ignore shared caching?

Author: Yes, this goes into how they use the `lifecycle`. 

1. They can omit providing a `-shared-cache-{dir.image}` and the lifecycle would store the cache data in the standard cache.
2. They can `/dev/null` equivalent the `-shared-cache-{dir}`.

> In `pack`, it being a local development tool, shared cached layers could be implemented as a single global cache that is shared across all applications. Additional options can be provided to restrict or otherwise configure this sharing.

Reviewer 3: ```suggestion
In `pack`, being a local development tool, shared cached layers could be implemented as a single global cache that is shared across all applications. Additional options can be provided to restrict or otherwise configure this sharing.
```

In a different platform, one can foresee that the cache is shared only within a "project" or an "organization".

#### Lifecycle Implementation

> 1. The lifecycle will provide additional `-shared-cache-dir` and `-shared-cache-image` options where `-cache-dir` and `-cache-image` could be provided.

Reviewer: I feel like "global" is better than "shared". But this might also be my bikeshed.

Reviewer 2: I think `global` implies it is shared will all builds but that may not be true depending on how the platform to implement sharing. It could be shared only within a subset of builds owned by a given user, for example. `pack` on the other hand could use the word `global` if that was truly the only scope for sharing.

Author: Ditto @ekcasey. In my mind, "global" explicitly means shared with everyone where "shared" leaves it up to interpretation.

2. Layers marked as `cache = true` and `cache-scope = "shared"` will be stored in either `-shared-cached-dir` or `-shared-cache-image`.

> 3. If `-share-cached-dir` or `-share-cached-image` are not provide and a buildpack sets a layer to cache with scope `shared`, then it would be cached to the standard cache location (`-cache-dir`/`-cache-image`).

Reviewer 3: ```suggestion
3. If `-share-cached-dir` or `-share-cached-image` are not provided and a buildpack sets a layer to cache with scope `shared`, then it would be cached to the standard cache location (`-cache-dir`/`-cache-image`).
```

> 3. If `-share-cached-dir` or `-share-cached-image` are not provide and a buildpack sets a layer to cache with scope `shared`, then it would be cached to the standard cache location (`-cache-dir`/`-cache-image`).

Reviewer 2: ```suggestion
3. If `-shared-cache-dir` or `-shared-cache-image` are not provided and a buildpack sets a layer to cache with scope `shared`, then it would be cached to the standard cache location (`-cache-dir`/`-cache-image`).
```

# Drawbacks

> [drawbacks]: #drawbacks

Reviewer: Are there any security concerns? Like, could there be secrets meant to be used on one buildpack (like a Maven `settings.xml`) that accidentally get shared with another less-trusted buildpack? It seems not that big of a problem since a buildpack can already do just about any thing (you shouldn't be using untrusted buildpacks).

Are there any debugging concerns? Could a buildpack with a bug that creates a malformed `.m2` buildpack cause an error in another buildpack? What if two buildpacks intend to share a cache layer, but they are using two different versions of some shared schema?

Reviewer 4: > What if two buildpacks intend to share a cache layer, but they are using two different versions of some shared schema?

By "two buildpacks" do you mean buildpacks with different `id`s? I'd interpreted the `This cached layer is specific to the buildpack` in the "what it is" section as meaning the cache is still buildpack-specific, in which case this scenario can't occur?

That said, I can see a few other problematic shared-cache scenarios:
1. The shared-cache is used by two different major versions of the same buildpack, where the cache contents is not compatible between them (for example a new version of the yarn buildpack defaulting to newer yarn, which has a different yarn cache schema)
2. The shared-cache contents is dependant on app source input (for example yarn version specified in `package.json`) meaning the shared-cache can be incompatible even with the same version of buildpack used across all apps

To help with (1), shared-caches could be made specific not only to the buildpack ID, but the buildpack version too. 

However:
* this wouldn't help with (2), so buildpack authors will still need to think about cache versioning themselves
* non-shared caches (presumably) aren't tied to buildpack version either, so doing this would be inconsistent

...so it making the cache buildpack-version specific doesn't seem worthwhile to me.

Though it does seem like it would be worth warning buildpack authors about general cache risks in the spec? (eg "Buildpacks SHOULD ensure that use of cached layers is forwards and backwards compatible across buildpack and tooling versions".)

> - Requires that the buildpack author be continuous about providing this information for optimization purposes.

Reviewer: I don't understand this sentence

Reviewer 3: Agreed; did you mean consistent?

Author: 🤦 `conscious`

Author: ```suggestion
- Requires that the buildpack author be conscious about providing this information for optimization purposes.
```

# Alternatives
[alternatives]: #alternatives

- Generic Read/Write volume mounts (RFC PR [#85](https://github.com/buildpacks/rfcs/pull/85))
    - The buildpack user must know what and where to mount things.
    - Has safety concerns when writing to Linux mounts.
    - Solution violates the spec dependent on use.
    - Solution is specific to `pack`.
    
# Prior Art
[prior-art]: #prior-art

None

# Unresolved Questions

> [unresolved-questions]: #unresolved-questions

Reviewer: If one buildpack need to wipe (`rm -rf`) a cache layer, does that mean other buildpacks will lose it too?

# Spec. Changes
[spec-changes]: #spec-changes

### Buildpacks Spec

##### [Layer Content Metadata TOML](https://github.com/buildpacks/spec/blob/main/buildpack.md#layer-content-metadata-toml)
```toml
# whether this layer should be available to at launch (run-time)
launch = false

# whether this layer should be available to subsequent buildpacks
# type: boolean
build = false

# whether this layer should be cached
# boolean
# type: boolean
cache = false

# type of cache (if `cache = true`)
# type: string
# values: app, shared
# default: app

> cache-scope = "app"

Reviewer: Maybe `local` instead of `app`? I feel like we should avoid "app"

Do we anticipate other values in the future? Why not maybe this a boolean `cache-shared`?

> [metadata]

Reviewer 2: I wonder if `metadata` gets problematic in this situation. The buildpack would need to be careful not to fight with itself about the contents of the metadata. 

E.G it would append new dependencies to the metadata if it adds deps to a shared layers (if it is recording that type of thing) instead of replacing the metadata wholesale with the list of deps relevant to a given build.

- CNB Issue: (leave blank)
- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)

# Summary
[summary]: #summary

Allow for buildpacks to specify what cache layers can be shared across applications.

# Motivation
[motivation]: #motivation

In order to provide a better experience, platforms such as `pack` could provide a way to share cache across multiple applications. In particular, this is a frequently requested feature ([pack#340](https://github.com/buildpacks/pack/issues/340), [pack#652](https://github.com/buildpacks/pack/issues/652), [slack](https://buildpacks.slack.com/archives/C94UJCNV6/p1588718688412300), [slack](https://buildpacks.slack.com/archives/CD61YAG69/p1586370336098700)) for similar language family apps.

Use Case:
```
Given I build a Java app A
When I build a Java app B
Then I expect similar maven dependencies not to have to be downloaded again
```

In order to solve this use case in a safe manner, the platform needs to know what cache it can share with other applications versus what cache is specific to a single application. This information is primarily known by the buildpacks.

# What it is
[what-it-is]: #what-it-is

> Buildpacks will provide a mechanism to denote a cached layer as something that can be shared by associating a "type" to the cache.

Reviewer 5: If I understand correctly, the buildpack author will decide what type of cache is going to be used, right?
I have a few questions about it:
1. Will the author be able to use both types of cache (each one for different purpose)?
2. If the buildpack author is using cache type x, does it make sense that the buildpack user won't want to use this type of cache? If yes, what will happen in this scenario? (I can't think about such real world scenario but I'm not familiar enough with our use cases)
3. What will happen if the buildpack author doesn't specify the cache type? I assume the default cache is `app-specific`, right?

Author: 1. Yes, the buildpack author would be able to set a cache type per cache layer.
2. This is an interesting question. This would ultimately be up to the platforms to determine what level of control they want to provide to the end user. As you mention, it's hard to think of a case where this would be necessary since this is ultimately just creating a distinction and not so much changing any functionality.
3. `app-specific` would be the default (I'll update the doc).
